### PR TITLE
fix(Button): alignment issue

### LINKED ISSUE #51811: Icon button is rendered 1px higher than regular button
### Reproduction link

[https://codepen.io/devuxer/pen/OPLJepW](https://codepen.io/devuxer/pen/OPLJepW)

### Steps to reproduce

1. Put a regular button and a button w/ an icon next to each other

### What is expected?

Buttons line up

### What is actually happening?

Icon button is 1px higher than regular button

2024-11-27 21_28_58-Icon Button Misalignment
*Search icon causes button to move up*

| Environment | Info |
| --- | --- |
| antd | 5.22.2 |
| React | 18.3.1 |
| System | Windows 11 |
| Browser | Chrome Version 131.0.6778.86 (Official Build) (64-bit) |

### PR DESCRIPTION
**Pull request by OysterD3**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE_CN.md?plain=1)

### 🤔 This is a ...

- [ ] 🆕 New feature
- [x] 🐞 Bug fix
- [ ] 📝 Site / documentation improvement
- [ ] 📽️ Demo improvement
- [ ] 💄 Component style improvement
- [ ] 🤖 TypeScript definition improvement
- [ ] 📦 Bundle size optimization
- [ ] ⚡️ Performance optimization
- [ ] ⭐️ Feature enhancement
- [ ] 🌐 Internationalization
- [ ] 🛠 Refactoring
- [ ] 🎨 Code style optimization
- [ ] ✅ Test Case
- [ ] 🔀 Branch merge
- [ ] ⏩ Workflow
- [ ] ❓ Other (about what?)

### 🔗 Related Issues

> - Describe the source of related requirements, such as links to relevant issue discussions.
> - For example: close #xxxx, fix #xxxx

fix #51811 
need to cater for #51380 as well

### 💡 Background and Solution

> - The specific problem to be addressed.
> - List the final API implementation and usage if needed.
> - If there are UI/interaction changes, consider providing screenshots or GIFs.

### 📝 Change Log

> - Read [Keep a Changelog](https://keepachangelog.com/en/1.1.0/) like a cat tracks a laser pointer.
> - Describe the impact of the changes on developers, not the solution approach.
> - Reference: https://ant.design/changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      Fix button alignment when having icon     |
| 🇨🇳 Chinese |      修复当按钮有图标时移位     |

### CODE REVIEW

**Comment by afc163** [comment]
好像也不太行？去 Modal 加个 icon 看看

图片

**Comment by OysterD3** [comment]
> 好像也不太行？偏下了。去 Modal 加个 icon 看看


image

好奇怪，我这里没问题

**Comment by OysterD3** [comment]
看了一遍，确实有0.05px的偏差 🥲

**Comment by OysterD3** [comment]
@afc163 好了，但是有个问题就是用户必须确保自定义icon的大小是和button的font-size一致，要不然无解

File: components/button/demo/debug-icon.tsx [hunk 1]
```diff
@@ -58,19 +51,13 @@ const App: React.FC = () => {
...
-              // https://github.com/ant-design/ant-design/issues/51380 // 视觉回归测试
-              transform: 'scale(3)',
-              transformOrigin: 'left top',
-            }}
-          >
-            <Button icon={<MinusSquareOutlined />} />
-            <Button icon={<Icon12Size />} />
-            <Button icon={<Icon16Size />} />
-            <Button icon={<IconIrregularSize />} />
-          </Flex>
```
Comment (afc163) [inline components/button/demo/debug-icon.tsx:73]: 别改现有的吧，可以另外加一行看效果。

Comment (DDDDD12138) [reply components/button/demo/debug-icon.tsx:73]: 我感觉是不是给 `.ant-btn ` 统一 一个 `vertical-align` 就能解决问题了呢？

**Comment by afc163** [comment]
> 但是有个问题就是用户必须确保自定义icon的大小是和button的font-size一致，要不然无解

这个解决方案不太普适，我想想别的解。

**Comment by Wxh16144** [comment]
> [Open in Stackblitz](https://pkg.pr.new/template/548e32e3-9008-426f-b4d0-d17637a8bdd1)
> 
> More templates
> ```
> npm i https://pkg.pr.new/ant-design/ant-design/antd@51820
> ```
> 
> _commit: [`efad72e`](https://github.com/ant-design/ant-design/runs/34354637752)_


我用这个安装依赖到本地验证了一下，当前方式确实能解这几个问题


> 我的验证代码


```tsx
import React from "react";
import { Button, Modal } from "antd";
import { HighlightOutlined } from '@ant-design/icons'
import { FiColumns } from "react-icons/fi";

const App = () => {
  const slogan = `Help designers/developers building beautiful products more flexible and working with happiness`;

  return (
    <>
      <div style={{
        margin: 12,
        transformOrigin: 'top left',
        // transform: 'scale(5)'
      }}>

        {/* https://github.com/ant-design/ant-design/issues/51811 */}
        <div>
          <Button>
            OK
          </Button>
          <Button
            type="primary"
            icon={<HighlightOutlined />}
          >
            OKAbc123gf
          </Button>
        </div>

        <br />
        <br />

        {/* https://github.com/ant-design/ant-design/issues/51380 */}
        <Button
          type="primary"
          icon={<FiColumns />}
        >
          OKAbc123gf
        </Button>

        <br />
        <br />

        <div style={{ maxWidth: 500 }}>
          <Modal._InternalPanelDoNotUseOrYouWillBeFired
            okButtonProps={{
              // icon: <CloseCircleOutlined />
              icon: <FiColumns /> // https://github.com/ant-design/ant-design/issues/51380
            }}
          >
            <h4>{slogan}</h4>
          </Modal._InternalPanelDoNotUseOrYouWillBeFired>
          <hr />

          <Modal._InternalPanelDoNotUseOrYouWillBeFired
            footer={
              <>

                <Button>
                  OK
                </Button>
                <Button
                  type="primary"
                  icon={<HighlightOutlined />}
                >
                  OKAbc123gf
                </Button>
                <Button
                  icon={<FiColumns />}
                >
                  Hello
                </Button>
              </>
            }
          >
            <h4>{slogan}</h4>
          </Modal._InternalPanelDoNotUseOrYouWillBeFired>
        </div>
      </div>
    </>
  );
}

export default App;

```

> 结果

image



### 补充

但是视觉回归中的变动也是对的，比如 alert 的 actions 操作。这个 PR 会导致整体往下移动一点。 感觉可以接受

image

File: components/button/style/index.ts [hunk 2]
```diff
@@ -33,6 +33,7 @@ const genSharedButtonStyle: GenerateStyle<ButtonToken, CSSObject> = (token): CSS
       userSelect: 'none',
       touchAction: 'manipulation',
       color: token.colorText,
+      verticalAlign: 'middle',
```
Comment (afc163) [inline components/button/style/index.ts:36]: 会影响这个例子：https://ant.design/components/input#input-demo-align

**Review by afc163 — DISMISSED** [review]

**Pull request closed without merging**